<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/hpdp/NoCarNoFriend/Assignment_6_Big_Data_Handling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 6: Big Data Handling


## Group Name: NO CAR NO FRIEND

### Group Members:
| Name                                      | Matrix Number   |
| :---------------------------------------- | :-------------: |
| SAM CHIA YUN                              |A21EC0127        |
| YEW RUI XIANG           |A21EC0149        |

## Pick a Big Dataset
In order to proceed for the Big Data Handling process, firstly we need to select a suitable dataset. We will be searching our dataset from the Kaggle Website since choosing appropriate one is crucial for the success of the project.

###Dataset Information
**Name :**  Public Employees Salaries From Brazil SP 2012<br>
**URL:** [*Dataset Link*](https://www.kaggle.com/datasets/bambrozim/public-employees-salaries-from-brazilsp-2012) <br>
**Description:** The "Public Employees Salaries From Brazil SP 2012" on Kaggle offers an aggregation of data public available on transparency portal from São Paulo, Brazil regarding the salaries from public employees. It includes features such as name, department, position, month total earnings, gross total earnings and net total earnings.


## Loading the Dataset

The code below is uploading your Kaggle API credentials, configuring the necessary directory structure, and then using the Kaggle API to download our dataset from the Kaggle which is Public Employees Salaries From Brazil SP 2012.

In [ ]:
# Upload kaggle.json API token, and download / unzip csv
from google.colab import files
files.upload()

!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json
!kaggle datasets download -d bambrozim/public-employees-salaries-from-brazilsp-2012

Saving kaggle.json to kaggle (3).json
mkdir: cannot create directory ‘/root/.kaggle’: File exists
100% 3.05G/3.07G [00:25<00:00, 154MB/s]
100% 3.07G/3.07G [00:25<00:00, 128MB/s]


In the code below, we define the names of the ZIP file and the specific file within the ZIP archive that we want to extract. Then we unzips the specific file from the ZIP archive, placing it in the current working directory so we can use it as our dataset later and finally removes the original ZIP file.

In [ ]:
import zipfile
import os

zip_file_name = 'public-employees-salaries-from-brazilsp-2012.zip'
initial_target_file = 'all_merged/2023_jan_august_merged.csv'  # Replace with the actual file name

# Unzip the file
with zipfile.ZipFile(zip_file_name, 'r') as zip_ref:
    # Extract the specific file to the current working directory
    zip_ref.extract(initial_target_file, '.')

# Remove the zip file after extracting
os.remove(zip_file_name)

Importing essential libraries for data analysis and visualization.

In [ ]:
# import required datasets
import numpy as np
import pandas as pd
import gc
import time
# import matplotlib.pyplot as plt

Measuring the execution time of reading a CSV file into a Pandas DataFrame using the %time magic command. Besides, the resulting df(DataFrame) will contain the data from the csv file that we had unzipped just now, so we can further analyze it.

In [ ]:
%time
df = pd.read_csv(initial_target_file, sep=';')

CPU times: user 2 µs, sys: 1 µs, total: 3 µs
Wall time: 7.39 µs


The head() method is then used to display the first few rows of the DataFrame, providing a quick overview of the dataset's structure and content.

In [ ]:
df.head()

NAME                        POSITION  \
0     AAIRON TELES DE CAMARGO          OFICIAL ADMINISTRATIVO   
1       AALAN DA SILVA ARAUJO                        SD 1C PM   
2  AARAN ESTEVAO LIMA BARBOSA                        CAP   PM   
3           AARAO DE OLIVEIRA                        MAJ   PM   
4     AARAO HENRIQUE OLIVEIRA  AG. ESCOLTA E VIGILANCIA PENIT   

                       DEPARTMENT MONTH_TOTAL VACATION_13TH_SALARY  \
0   DEPT.ESTAD.TRANSITO DETRAN SP     2645,13                 0,00   
1       POLICIA MILITAR SAO PAULO     4469,53                 0,00   
2       POLICIA MILITAR SAO PAULO    13986,90                 0,00   
3  SAO PAULO PREVIDENCIA   SPPREV    17921,51                 0,00   
4  SECRET.ADMINISTR.PENITENCIARIA     3717,47                 0,00   

  EVENTUAL PAYMENTS PAYED_PREMIUM_LEAVE OTHER_INDEMNITIES WAGE_REDUCER  \
0              0,00                0,00            193,29         0,00   
1              0,00                0,00           1027,80         0,00   
2              0,00                0,00            616,68         0,00   
3              0,00                0,00              0,00         0,00   
4              0,00                0,00            186,34         0,00   

  NET_TOTAL       STATUS  MONTH  YEAR  
0   1167,57       ATIVOS  APRIL  2023  
1   2678,84       ATIVOS  APRIL  2023  
2   7592,82       ATIVOS  APRIL  2023  
3  16574,69  APOSENTADOS  APRIL  2023  
4   2753,87       ATIVOS  APRIL  2023

In order to check the total number of rows and columns of the dataframe, we can use df.shape

In [ ]:
df.shape

(8533848, 13)

We use df.info() to quickly assessing the size of the DataFrame and the types of data it contains.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8533848 entries, 0 to 8533847
Data columns (total 13 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   NAME                  object
 1   POSITION              object
 2   DEPARTMENT            object
 3   MONTH_TOTAL           object
 4   VACATION_13TH_SALARY  object
 5   EVENTUAL PAYMENTS     object
 6   PAYED_PREMIUM_LEAVE   object
 7   OTHER_INDEMNITIES     object
 8   WAGE_REDUCER          object
 9   NET_TOTAL             object
 10  STATUS                object
 11  MONTH                 object
 12  YEAR                  int64 
dtypes: int64(1), object(12)
memory usage: 846.4+ MB


Iterating over the columns of a DataFrame, prints information about each column, specifically the column index, name, and the number of unique values in that column.  This can be useful for us to understand the diversity of values in each column of the DataFrame.

In [ ]:
i = 0
for column in df.columns:
  print("Number of unique values in [" + str(i) + "] column (" + column + "): " + str(len(df[column].unique())))
  i += 1

Number of unique values in [0] column (NAME): 1008824
Number of unique values in [1] column (POSITION): 3323
Number of unique values in [2] column (DEPARTMENT): 104
Number of unique values in [3] column (MONTH_TOTAL): 909657
Number of unique values in [4] column (VACATION_13TH_SALARY): 394138
Number of unique values in [5] column (EVENTUAL PAYMENTS): 298536
Number of unique values in [6] column (PAYED_PREMIUM_LEAVE): 14966
Number of unique values in [7] column (OTHER_INDEMNITIES): 153007
Number of unique values in [8] column (WAGE_REDUCER): 26191
Number of unique values in [9] column (NET_TOTAL): 1303509
Number of unique values in [10] column (STATUS): 3
Number of unique values in [11] column (MONTH): 8
Number of unique values in [12] column (YEAR): 1


Showing the columns that we have in the DataFrame.

In [ ]:
df.columns

Index(['NAME', 'POSITION', 'DEPARTMENT', 'MONTH_TOTAL', 'VACATION_13TH_SALARY',
       'EVENTUAL PAYMENTS', 'PAYED_PREMIUM_LEAVE', 'OTHER_INDEMNITIES',
       'WAGE_REDUCER', 'NET_TOTAL', 'STATUS', 'MONTH', 'YEAR'],
      dtype='object')

Calculating the initial memory usage of a Pandas DataFrame (df) in megabytes. We perform this memory calculation before any significant operations are performed on the DataFrame so that we could make a comparison later after applying different strategies to the dataset.

In [ ]:
initial_memory_usage = df.memory_usage(deep=True).sum() / (1024 * 1024)
initial_memory_usage

6644.939455032349

In [ ]:
# Clear memory of dataframe
del df

## Strategies for Big Datasets

We will be applying five smart strategies to handle the large dataset effectively.
* Load Less Data
* Optimize Data Types
* Use Chunking
* Sampling
* Parallelize with Dask

### 1. Load Less Data

When dealing with large datasets, it's often unnecessary to load the entire dataset into memory if we only need a subset for analysis. We need to utilize features in our data loading library that allow us to load specific columns or rows of interest. This strategy significantly reduces memory usage and speeds up data loading times.

When processing data, it's advisable to avoid loading unnecessary columns. For instance, in this dataset focused on the year 2023, there's no need to include the "YEAR" column in the dataframe. Opt for loading only the variables essential for analysis and training purposes.

In [ ]:
# Load only required columns when read data
columns_needed = ['NAME', 'POSITION', 'DEPARTMENT', 'MONTH_TOTAL', 'VACATION_13TH_SALARY',
       'EVENTUAL PAYMENTS', 'PAYED_PREMIUM_LEAVE', 'OTHER_INDEMNITIES',
       'WAGE_REDUCER', 'NET_TOTAL', 'STATUS', 'MONTH']

df = pd.read_csv(initial_target_file, sep=';', usecols=columns_needed)

Calculating the memory usage of the DataFrame (df) in megabytes after we load lesser data. As the result below showing the memory usage is only 6579.83 MB after we avoid loading unnecessary columns compared to the initial memory usage which is 6644.94 MB.

In [ ]:
memory_usage_1 = df.memory_usage(deep=True).sum() / (1024 * 1024)
memory_usage_1

6579.83136177063

Since the original dataset is too large to be fully loaded, therefore we will only loads a subset of the dataset which we limit it to 100000 rows and only load the neccessary columns. This allows us for a quick initial exploration and insight into the data while conserving memory.

The summary of the DataFrame structure and memory usage is also showed with the code below. Therefore, we only use 77.0 MB of memory and get a rapid overview of the dataset.

In [ ]:
# Load less data by specifying nrows when read data
# Useful for extracting an initial insight about data
df_nrow = pd.read_csv(initial_target_file, sep=';', usecols=columns_needed, nrows=100000)
df_nrow.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 12 entries, NAME to MONTH
dtypes: object(12)
memory usage: 77.0 MB


### 2. Optimize Data Types

Efficiently managing data types can significantly reduce memory usage. By selecting the most space-efficient data types for the columns, we optimize memory usage without compromising data integrity.

This code below is converting the selected columns: 'POSITION', 'DEPARTMENT', 'STATUS' and 'MONTH' to categorical data type. This could optimize memory usage for categorical variables as it can lead to significant memory savings while preserving the analytical capabilities of the variables.

In [ ]:
# Shrink categorical data using Categorical dtypes
columns_cat = ['POSITION', 'DEPARTMENT', 'STATUS', 'MONTH']
df[columns_cat] = df[columns_cat].astype('category')

Providing a summarized view of the DataFrame's structure and its memory usage. We apply 'verbose=False' and 'memory_usage="deep"' is to make the output more concise and memory-aware.

In [ ]:
df.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8533848 entries, 0 to 8533847
Columns: 12 entries, NAME to MONTH
dtypes: category(4), object(8)
memory usage: 4.1 GB



By executing the code below, the numeric columns in the DataFrame will have commas removed from their values. This is because for Brazil Currency they will be using ' , ' instead of ' . ' to represent the decimal points, so to continue with our data analysis we would like to remove the ',' from the data. Integer type always taking less memory space compared to float type.

In [ ]:
# Data Cleaning to remove unwanted character
columns_num = ['MONTH_TOTAL', 'VACATION_13TH_SALARY',
       'EVENTUAL PAYMENTS', 'PAYED_PREMIUM_LEAVE', 'OTHER_INDEMNITIES',
       'WAGE_REDUCER', 'NET_TOTAL']

for column in columns_num:
  df[column] = df[column].str.replace(',', '')

Displaying the first few rows of the DataFrame, providing a quick overview of the dataset's structure and content.

In [ ]:
df.head()

NAME                        POSITION  \
0     AAIRON TELES DE CAMARGO          OFICIAL ADMINISTRATIVO   
1       AALAN DA SILVA ARAUJO                        SD 1C PM   
2  AARAN ESTEVAO LIMA BARBOSA                        CAP   PM   
3           AARAO DE OLIVEIRA                        MAJ   PM   
4     AARAO HENRIQUE OLIVEIRA  AG. ESCOLTA E VIGILANCIA PENIT   

                       DEPARTMENT MONTH_TOTAL VACATION_13TH_SALARY  \
0   DEPT.ESTAD.TRANSITO DETRAN SP      264513                  000   
1       POLICIA MILITAR SAO PAULO      446953                  000   
2       POLICIA MILITAR SAO PAULO     1398690                  000   
3  SAO PAULO PREVIDENCIA   SPPREV     1792151                  000   
4  SECRET.ADMINISTR.PENITENCIARIA      371747                  000   

  EVENTUAL PAYMENTS PAYED_PREMIUM_LEAVE OTHER_INDEMNITIES WAGE_REDUCER  \
0               000                 000             19329          000   
1               000                 000            102780          000   
2               000                 000             61668          000   
3               000                 000               000          000   
4               000                 000             18634          000   

  NET_TOTAL       STATUS  MONTH  
0    116757       ATIVOS  APRIL  
1    267884       ATIVOS  APRIL  
2    759282       ATIVOS  APRIL  
3   1657469  APOSENTADOS  APRIL  
4    275387       ATIVOS  APRIL

The function 'optimize_data_types' optimizes the data types of numeric columns in the DataFrame to use the smallest integer types that can accommodate the range of values in each column. This process can reduce the memory footprint of the DataFrame. Besides, this function will also provide feedback on the initial and final memory usage, along with the percentage decrease achieved through the optimization. By comparing the initial and final memory usage we can have a clearly look on how the optimzation process reduce the memory usage of the DataFrame.

In [ ]:
def optimize_data_types(df):
  # Calculate initial memory of the dataframe
  start_mem = df.memory_usage().sum() / 1024**2
  print('Memory usage of dataframe is {:.2f} MB'.format(start_mem))

  # Change the numeric columns into integer datatype
  for col in columns_num:
    df[col] = df[col].astype('int64')
    c_min = df[col].min()
    c_max = df[col].max()

    if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
      df[col] = df[col].astype('int8')
    elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
      df[col] = df[col].astype('int16')
    elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
      df[col] = df[col].astype('int32')
    elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
      df[col] = df[col].astype('int64')

  # Calculate memory usage after optimization
  end_mem = df.memory_usage().sum() / 1024**2
  print('Memory usage after optimization is: {:.2f} MB'.format(end_mem))
  print('Decreased by {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem))

optimize_data_types(df)

Memory usage of dataframe is 561.71 MB
Memory usage after optimization is: 333.84 MB
Decreased by 40.6%


Calculating the memory usage of the DataFrame (df) in megabytes after we Optimize Data Types. As the result below showing the memory usage is only 925.37 MB after the optimization process compared to the initial memory usage which is 6644.94 MB.

In [ ]:
memory_usage_2 = df.memory_usage(deep=True).sum() / (1024 * 1024)
memory_usage_2

925.367600440979

Using df.info() to quickly assessing the size of the DataFrame and the types of data it contains.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8533848 entries, 0 to 8533847
Data columns (total 12 columns):
 #   Column                Dtype   
---  ------                -----   
 0   NAME                  object  
 1   POSITION              category
 2   DEPARTMENT            category
 3   MONTH_TOTAL           int32   
 4   VACATION_13TH_SALARY  int32   
 5   EVENTUAL PAYMENTS     int32   
 6   PAYED_PREMIUM_LEAVE   int32   
 7   OTHER_INDEMNITIES     int32   
 8   WAGE_REDUCER          int32   
 9   NET_TOTAL             int32   
 10  STATUS                category
 11  MONTH                 category
dtypes: category(4), int32(7), object(1)
memory usage: 333.8+ MB


Displaying the first few rows of the DataFrame, providing a quick overview of the dataset's structure and content.

In [ ]:
df.head()

NAME                        POSITION  \
0     AAIRON TELES DE CAMARGO          OFICIAL ADMINISTRATIVO   
1       AALAN DA SILVA ARAUJO                        SD 1C PM   
2  AARAN ESTEVAO LIMA BARBOSA                        CAP   PM   
3           AARAO DE OLIVEIRA                        MAJ   PM   
4     AARAO HENRIQUE OLIVEIRA  AG. ESCOLTA E VIGILANCIA PENIT   

                       DEPARTMENT  MONTH_TOTAL  VACATION_13TH_SALARY  \
0   DEPT.ESTAD.TRANSITO DETRAN SP       264513                     0   
1       POLICIA MILITAR SAO PAULO       446953                     0   
2       POLICIA MILITAR SAO PAULO      1398690                     0   
3  SAO PAULO PREVIDENCIA   SPPREV      1792151                     0   
4  SECRET.ADMINISTR.PENITENCIARIA       371747                     0   

   EVENTUAL PAYMENTS  PAYED_PREMIUM_LEAVE  OTHER_INDEMNITIES  WAGE_REDUCER  \
0                  0                    0              19329             0   
1                  0                    0             102780             0   
2                  0                    0              61668             0   
3                  0                    0                  0             0   
4                  0                    0              18634             0   

   NET_TOTAL       STATUS  MONTH  
0     116757       ATIVOS  APRIL  
1     267884       ATIVOS  APRIL  
2     759282       ATIVOS  APRIL  
3    1657469  APOSENTADOS  APRIL  
4     275387       ATIVOS  APRIL

Cleaned dataset is saved so we can have a comparation with the initial dataset afterwards.

Exporting the DataFrame to a CSV file. The cleaned dataset is saved so that we can make comparison with the initial dataset afterwards.

In [ ]:
df.to_csv('all_merged/small_size_2023_jan_august_merged.csv')

Deleting the DataFrame.

In [ ]:
del df

gc.collect() is a call to the garbage collector in Python. This function manually triggers the garbage collector to attempt to clean up unreferenced and unused objects in memory, thus improving the memory efficiency.

In [ ]:
gc.collect()

0

### 3. Use Chunking

Dividing the dataset into smaller, manageable chunks and processing them one at a time. This is particularly useful when the entire dataset cannot fit into memory.  By processing data in smaller pieces, we can perform computations iteratively, avoiding memory overflow issues.

**Traditional method**

####What is the average net total income for each status?

In order to answer this question, we load only 2 columns which 'NET TOTAL' and 'STATUS'.

There are 3 rows contained in the 'STATUS' column which are:
* APOSENTADOS :- RETIREES
* ATIVOS :- ACTIVE
* PENSIONISTAS :- PENSIONERS

In [ ]:
target_file = 'all_merged/small_size_2023_jan_august_merged.csv'
columns_needed = ['NET_TOTAL', 'STATUS']

We use '%%time' magic command to display the time taken for the execution of the entire cell. This command can conclude the performance of reading data from the CSV file and performing a grouping and aggregation operation.

In [ ]:
%%time
df = pd.read_csv(target_file, usecols = columns_needed)
mean = df.groupby('STATUS').agg('mean')

CPU times: user 9.9 s, sys: 645 ms, total: 10.5 s
Wall time: 10.6 s


Calculation the mean of NET_TOTAL for every different STATUS.

In [ ]:
mean

NET_TOTAL
STATUS                     
APOSENTADOS   522020.002970
ATIVOS        494071.963205
PENSIONISTAS  484023.680335

Deleting the dataframe.

In [ ]:
del df

**Chunking method**

We will be using '%%time' magic command to measures the time taken to run the entire cell. The code below will be reading the dataset in manageable chunks, then performing operations on each chunk and aggregating the results. This approach is memory-efficient and suitable when the entire dataset is too large to fit into the memory.

From the result below, we can see that the time taken to run the entire while using Chunking strategy and the initial processing time.

|      Scenario               | User  |   Sys   |  Total  |  Wall time |
| :-------------------- | :-------------: |  :-------------: |  :-------------: |  :-------------: |
| Initial processing time      | 2 µs         | 1 µs  | 3 µs | 7.39 µs |
| Using Chunking Strategy           |10 s        | 426 ms| 10.5 s| 10.5 s|

It is clear that chunking method will take longer execution time when we compare the processing time of chunking and the initial processing time.

In [ ]:
%%time
# Initialize an empty DataFrame to store the aggregated results
aggregated_results = pd.DataFrame()

# Iterate through chunks of the CSV file
for chunk in pd.read_csv(target_file,  usecols = columns_needed, chunksize=500000):
    # Perform your operations on each chunk
    mean_chunk = chunk.groupby('STATUS').agg('mean')

    # Append the aggregated results for the chunk to the final DataFrame
    aggregated_results = aggregated_results.append(mean_chunk)

# Calculate the overall mean from the aggregated results
final_mean = aggregated_results.groupby('STATUS').agg('mean')

<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from panda

CPU times: user 10 s, sys: 426 ms, total: 10.5 s
Wall time: 10.5 s


<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
<timed exec>:10: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.


Showing the mean of NET_TOTAL for every different STATUS.

From the results we can see that:
* The mean of 'Retirees' Status rised up from 522020.00 to 524646.91
* The mean of 'Active' Status dropped down from 494071.96 to 493732.53
* The mean of 'PENSIONERS' rised up from 484023.68 to 484683.35

The result is almost similar to the result that we produced by using the original DataFrame, so it will not cause any bad effects to our result.


In [ ]:
final_mean

NET_TOTAL
STATUS                     
APOSENTADOS   524646.906214
ATIVOS        493732.528653
PENSIONISTAS  484683.353709

### 4. Sampling

Instead of examining the entire dataset, extracting a representative sample can provide valuable insights, especially when the dataset is too extensive for comprehensive analysis. Employing methods such as random sampling or systematic sampling enables a more manageable exploration. Analyzing a subset of the data facilitates rapid prototyping, hypothesis testing, and initial exploration.



We apply the '%%time' magic command to measures the time taken to run the entire cell. For the code below, we read the dataset into chunks then applies a sampling method to each chunk, concatenates the sampled data into single DataFrame. After that, we will calculates the mean for each group in 'STATUS' column so that we can have a comparison between the results of applying different strategy.

From the result below, we can see that the time taken to run the entire while using Sampling strategy and the initial processing time.

|      Scenario               | User  |   Sys   |  Total  |  Wall time |
| :-------------------- | :-------------: |  :-------------: |  :-------------: |  :-------------: |
| Initial processing time      | 2 µs         | 1 µs  | 3 µs | 7.39 µs |
| Using Sampling Strategy | 30 µs | 0 ns | 30 µs | 33.9 µs |

It is clear that Chunking strategy has the ability to handle large datasets but will only cause a tiny increase in the processing time.



In [ ]:
%time
# Use sampling method to read the dataset for furture processing
df = pd.DataFrame()

for chunk in pd.read_csv(target_file,  usecols = columns_needed, chunksize=500000):
  # Select random set of data from the chunk with specified fraction which is 0.25
  chunk = chunk.sample(frac=0.25, random_state=99)
  # Concatenate the dataframe from each chunk
  df = pd.concat([df, chunk], ignore_index=True)

mean = df.groupby('STATUS').agg('mean')

CPU times: user 30 µs, sys: 0 ns, total: 30 µs
Wall time: 33.9 µs


We use df.shape to check the number of rows and columns of the sampled DataFrame.

In [ ]:
df.shape

(2133462, 2)





Displaying the mean of NET_TOTAL for every different STATUS.

From the result below, we can conclude that:-
* The mean of 'Retirees' Status dropped down from  524646.91 to 522601.79
* The mean of 'Active' Status dropped down from 493732.53 to 493434.15
* The mean of 'PENSIONERS' dropped down from 484683.35 to 482984.75

Therefore, by comparing the mean value that we calculated, we can say that sampling can represent the original dataset but producing higher processing speed compared to using the original dataset.

In [ ]:
mean

NET_TOTAL
STATUS                     
APOSENTADOS   522601.791360
ATIVOS        493934.150235
PENSIONISTAS  482984.745128

Providing a informative summary of the DataFrame, emphasizing memory usage details and omitting the column information. This can help us assessing the memory footprint of the DataFrame.

In [ ]:
df.info(verbose=False, memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2133462 entries, 0 to 2133461
Columns: 2 entries, NET_TOTAL to STATUS
dtypes: int64(1), object(1)
memory usage: 149.7 MB


Deleting the DataFrame.

In [ ]:
del df

### 5. Parallelize with Dask

Dask is a parallel computing library that integrates seamlessly with Pandas and extends its capabilities for larger-than-memory datasets. Dask operates by breaking down computations into smaller tasks and parallelizing their execution across multiple cores or even distributed computing clusters. By leveraging Dask, we can perform operations on large datasets that would be impractical with traditional Pandas.

Importing the Dask DataFrame module and aliasing it as dd. So that we can use the Dask library's functionalities.

In [ ]:
import dask.dataframe as dd

Applying the '%%time' magic command to measures the time taken to run the entire cell. In the code below, we leverages Dask to read the dataset, perform a grouping and aggregation operation then compare the final result produced.

From the result below, we can see that the time taken to run the entire while using Parallelize with Dask strategy and the initial processing time.

|      Scenario               | User  |   Sys   |  Total  |  Wall time |
| :-------------------- | :-------------: |  :-------------: |  :-------------: |  :-------------: |
| Initial processing time      | 2 µs         | 1 µs  | 3 µs | 7.39 µs |
| Parelleize with Dask | 3 µs | 0 ns | 3 µs | 7.87 µs |

We can see that there is only a tiny difference between the processing time between using Dask strategy and the initial processing time. Therefore, it is clear that Dask is effective and great while handling huge dataset.

In [ ]:
%time
ddf = dd.read_csv(target_file,  usecols = columns_needed)
mean = ddf.groupby('STATUS').agg('mean')

result = mean.compute()

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 7.87 µs


Determining the number of partitions in the Dask DataFrame. We need to understand the number of partitions as it is essential for optimizing the performance of Dask computations.

In [ ]:
ddf.npartitions

17

Printing the result mean of 'NET_TOTAL' for different 'STATUS'.
From the results produced below, we can conclude that:-

* The mean of 'Retirees' Status dropped down from 522601.79 to 494071.96
* The mean of 'Active' Status rised up from 493732.53 to 522020.01
* The mean of 'PENSIONERS' dropped down from 484683.35 to 484023.68


In [ ]:
result

NET_TOTAL
STATUS                     
ATIVOS        494071.963205
APOSENTADOS   522020.002970
PENSIONISTAS  484023.680335

Deleting the Dask DataFrame.

In [ ]:
del ddf

## Comparative Analysis

### 1. Memory Usage
First of all, we will make a comparison between the memory usage.

The code below is providing a summary view of the memory usage of DataFrame in different scenarios:-
* Initial memory usage
* The memory usage after loading less data
* The memory usage after optimizing data types

From the results below, we can clearly see that initial memory usage is 6644.94 MB, the memory usage after loading less data is 6579.83 MB and the memory usage after optimizing data types is 925.37 MB.
Therefore, we can conclude that the there is a huge impact of strategies on the memory footprint of the DataFrame.

Apart from that, we also claculates the reduction in the size of the DataFrame after optimizing data types to the initial memory usage. The result showed is 86.07% of reducing in size, so it is clear that the strategy that we applied to the DataFrame had worked well.

In [ ]:
print(f"The size of dataframe is {initial_memory_usage:.2f} MB.")
print(f"The size of dataframe with load only required columns is {memory_usage_1:.2f} MB.")
print(f"The size of dataframe with load only required columns and optimize data types is {memory_usage_2:.2f} MB.")

print("\nPercentage of reducing in size of dataframe by {:.2f}%".format(100 * (initial_memory_usage - memory_usage_2) / initial_memory_usage))

The size of dataframe is 6644.939455032349 MB.
The size of dataframe with load only required columns is 6579.83136177063 MB.
The size of dataframe with load only required columns and optimize data types is 925.367600440979 MB.

Percentage of reducing in size of dataframe by 86.07%


### 2. Processing time
Apart from that, we will be comparing the processing time taken to run the entire cell for different scenarios.


|      Scenario               | User  |   Sys   |  Total  |  Wall time |
| :-------------------- | :-------------: |  :-------------: |  :-------------: |  :-------------: |
| Initial processing time      | 2 µs         | 1 µs  | 3 µs | 7.39 µs |
| Using Chunking Strategy           |10 s        | 426 ms| 10.5 s| 10.5 s|
| Using Sampling Strategy | 30 µs | 0 ns | 30 µs | 33.9 µs |
| Parelleize with Dask | 3 µs | 0 ns | 3 µs | 7.87 µs |

* The initial processing time is extremely short representing a quick operation.
* Chunking, while providing the ability to handle large datasets, introduces a significant increase in processing time, likely due to the overhead of processing data in smaller chunks.
* Sampling proves to be an extremely fast strategy, making it suitable for quick analysis or exploration of a subset of the data.
* Applying parallelization with Dask shows efficient processing, with low wall time and user time, indicating that Dask is effectively utilizing parallel and distributed computing capabilities.

In summary, the choice of strategy depends on the specific use case. Chunking might be suitable for handling large datasets efficiently, while Sampling and Dask Parallelization are effective for quick analysis and distributed computing, respectively.

### 3. File Size
Besides, we will be comparing the size of the file.

Size of Initial File:
* File Size: 1159.66 MB
* This is the size of the original dataset in its raw form.

Size of Optimized File:
* File Size: 1046.72 MB
* This is the size of the file after applying optimizations, including selecting only the required columns and optimizing data types.

In conclusion, the optimization strategies, which include selecting required columns and optimizing data types, have effectively reduced the size of the dataset, leading to a smaller memory footprint. After doing calculation, the strategies had successfully made a  reduction in size, with a percentage reduction of 9.75% compared to the original file size.




In [ ]:
def get_file_size(file_path):
    try:
        size = os.path.getsize(file_path)
        return size
    except FileNotFoundError:
        print(f"The file '{file_path}' does not exist.")
        return None

# Example usage:
initial_file_size = get_file_size(initial_target_file)
after_file_size = get_file_size(target_file)

if initial_file_size is not None:
    print(f"The size of initial file '{initial_target_file}' is {initial_file_size / (1024 * 1024):.2f} MB.")
if after_file_size is not None:
    print(f"The size of file after selection of required columns and optimize data types '{target_file}' is {after_file_size / (1024 * 1024):.2f} MB.")

The size of initial file 'all_merged/2023_jan_august_merged.csv' is 1159.6561679840088 MB.
The size of file after selection of required columns and optimize data types 'all_merged/small_size_2023_jan_august_merged.csv' is 1046.7162189483643 MB.


## Conclusion

Strategies:
* Load Less Data: By loading only the required columns for analysis, unnecessary data was excluded, resulting in a smaller memory footprint.

* Data Type Optimization: Adjusting data types to more memory-efficient representations further reduced the memory consumption of the dataset.

* Sampling: Utilizing a sampling strategy allowed for quick analysis and exploration of a representative subset of the data.

* Dask Parallelization: Leveraging Dask for parallel and distributed computing enhanced the efficiency of handling larger-than-memory datasets.

Size Reduction:
* The size of the dataset was reduced from 1159.66 MB in its original form to 1046.72 MB after applying optimization strategies. This represents a significant reduction of approximately 9.75%.

Impact on Handling Big Data:
* Memory Efficiency: The optimization strategies significantly improved memory efficiency by reducing the memory footprint of the dataset. This is crucial when dealing with large datasets that may not fit into memory.


* Processing Efficiency: Different strategies, such as chunking, sampling, and Dask parallelization, were employed to enhance processing efficiency. These strategies allow for the parallel processing of data, reducing overall processing time.


In a nutshell, employing a combination of optimization strategies tailored to the characteristics of the dataset enhances the ability to handle big data efficiently. Memory and processing efficiency are crucial considerations, and the trade-offs between these factors depend on the specific requirements of the analysis.